File: sdks/typescript/src/plugin/framing.ts
/** Length-prefixed JSON codec for plugin protocol version 1. */

import type { Socket } from "node:net";

import { AbyssPluginError } from "./errors.js";

const FRAME_HEADER_BYTES = 4;
const MAX_JSON_FRAME_BYTES = 16 * 1024 * 1024;

export async function writeJsonFrame(
  socket: Socket,
  value: unknown,
): Promise<void> {
  const payload = Buffer.from(JSON.stringify(value), "utf8");
  if (payload.length > MAX_JSON_FRAME_BYTES) {
    throw new AbyssPluginError(
      `plugin frame payload length ${payload.length} exceeds maximum ${MAX_JSON_FRAME_BYTES}`,
    );
  }
  const header = Buffer.allocUnsafe(FRAME_HEADER_BYTES);
  header.writeUInt32BE(payload.length);
  await new Promise<void>((resolve, reject) => {
    socket.write(Buffer.concat([header, payload]), (error) => {
      if (error === null || error === undefined) {
        resolve();
      } else {
        reject(error);
      }
    });
  });
}

export async function* readJsonFrames(socket: Socket): AsyncGenerator<unknown> {
  let buffered: Buffer<ArrayBufferLike> = Buffer.alloc(0);
  for await (const rawChunk of socket) {
    const chunk = Buffer.isBuffer(rawChunk)
      ? rawChunk
      : Buffer.from(rawChunk as Uint8Array);
    buffered = buffered.length === 0 ? chunk : Buffer.concat([buffered, chunk]);
    while (buffered.length >= FRAME_HEADER_BYTES) {
      const payloadLength = buffered.readUInt32BE(0);
      if (payloadLength > MAX_JSON_FRAME_BYTES) {
        throw new AbyssPluginError(
          `plugin frame payload length ${payloadLength} exceeds maximum ${MAX_JSON_FRAME_BYTES}`,
        );
      }
      const frameLength = FRAME_HEADER_BYTES + payloadLength;
      if (buffered.length < frameLength) {
        break;
      }
      const payload = buffered.subarray(FRAME_HEADER_BYTES, frameLength);
      buffered = buffered.subarray(frameLength);
      try {
        yield JSON.parse(payload.toString("utf8")) as unknown;
      } catch (error) {
        throw new AbyssPluginError("decode broker plugin JSON frame", {
          cause: error,
        });
      }
    }
  }
  if (buffered.length !== 0) {
    throw new AbyssPluginError("broker plugin stream ended within a frame");
  }
}

File: sdks/python/abyss_sdk/plugin/framing.py
"""Length-prefixed JSON codec for plugin protocol version 1."""

import json
import struct
from typing import Any, Optional

from .errors import AbyssPluginError
from .transport import PluginTransport

MAX_JSON_FRAME_BYTES = 16 * 1024 * 1024


def write_json_frame(stream: PluginTransport, value: object) -> None:
    payload = json.dumps(value, separators=(",", ":")).encode("utf-8")
    if len(payload) > MAX_JSON_FRAME_BYTES:
        raise AbyssPluginError(
            f"plugin frame payload length {len(payload)} exceeds maximum {MAX_JSON_FRAME_BYTES}"
        )
    stream.write_all(struct.pack(">I", len(payload)) + payload)


def read_json_frame(stream: PluginTransport) -> Optional[Any]:
    header = _read_exact(stream, 4, allow_initial_eof=True)
    if header is None:
        return None
    payload_length = struct.unpack(">I", header)[0]
    if payload_length > MAX_JSON_FRAME_BYTES:
        raise AbyssPluginError(
            f"plugin frame payload length {payload_length} exceeds maximum {MAX_JSON_FRAME_BYTES}"
        )
    payload = _read_exact(stream, payload_length, allow_initial_eof=False)
    assert payload is not None
    try:
        return json.loads(payload)
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise AbyssPluginError("decode broker plugin JSON frame") from error


def _read_exact(stream: PluginTransport, size: int, *, allow_initial_eof: bool) -> Optional[bytes]:
    chunks = bytearray()
    while len(chunks) < size:
        chunk = stream.read(size - len(chunks))
        if not chunk:
            if allow_initial_eof and not chunks:
                return None
            raise AbyssPluginError("broker plugin stream ended within a frame")
        chunks.extend(chunk)
    return bytes(chunks)
